src/olympia/files/models.py

```diff
@@ -590,8 +590,17 @@ def add_file(self, chunks, filename, size):
         filename = smart_str(u'{0}_{1}'.format(self.uuid, filename))
         loc = os.path.join(user_media_path('addons'), 'temp', uuid.uuid4().hex)
         base, ext = os.path.splitext(smart_path(filename))
+
+        # Change a ZIP to an XPI, to maintain backward compatibility
+        # with older versions of Firefox and to keep the rest of the XPI code
+        # path as consistent as possible for ZIP uploads.
+        # See: https://github.com/mozilla/addons-server/pull/2785
+        if ext == '.zip':
+            ext = '.xpi'
+
```

Comment:
Testing this, I believe https://github.com/mozilla/addons-server/blob/master/src/olympia/files/tests/test_models.py#L563 is the correct way to go. Even better, https://github.com/mozilla/addons-server/blob/master/src/olympia/files/tests/test_models.py#L866 - `FileUpload.from_post` calls directly to `add_file` so that'll be fine.